text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+the disk space consumed by the databases, since the migrated data cannot be allowed to overwrite the existing data while
+the node is still running as a validator, and _this_ disk consumption might prove to be prohibitive.  Data migration
+would also still need to be executed again once the old `casper-node` stops due to reaching the upgrade activation
+point, since more data will likely have been written to the databases between then and when the data migration was
+previously completed.
+
+### Config migration
+
+Migration will also need to handle converting an old config file to a new one.  For this to work, we need to know which
+config values have been modified by the user from their default values.  This requires either default values to be
+provided, hard-coded in the node, or else the `migrate-data` subcommand will need to be passed not just the old and new
+config.toml files, but also a copy of each with their default values set.
+
+We moved away from hard-coding defaults in the code a while ago, but a challenge to the other approach is ensuring the
+user _copies_ the installed vanilla config rather than renaming or deleting it.  Two possible solutions are:
+1. have the installer create a read-only copy of the default config exclusively for the `upgrade-bootstrapper`'s use,
+   and a separate writable one for the user to modify/copy/move as they wish
+1. require that every config value is annotated with a standard comment specifying the default value
+
+It seems like annotating the config file with default values might be somewhat more "normal" and should be fairly simple
+to implement and test.
+
+To migrate, the node will identify any modified values in the old config, and apply the same changes to the new config
+if the corresponding new config value has not already been modified by the user.
+
+
+### `upgrade-bootstrapper`
+
+We also need to provide an `upgrade-bootstrapper` binary.  It should live in its own repo outside the `casper-node` one
+if setting up CI is not deemed too onerous.  Ideally its tests shouldn't need to run when changes are made to
+`casper-node`, and vice-versa.  There also should be no direct dependency between the `update-bootstrapper` and any of
+the other casper crates.  Hence it makes sense to keep this separate to the rest of the node project.
+
+The `upgrade-bootstrapper` and `casper-node` must both be capable of being restarted, having been stopped or killed at
+any step above, without disrupting this normal flow.  This will require the `upgrade-bootstrapper` to write which step
+its about to perform to the hard-coded known location immediately before performing the step.  On starting, the
+`upgrade-bootstrapper` will try to read this value and jump to that step.  If the value/file is absent, it assumes it's
+a fresh start.
+
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+Users lose the ability to override config values via command line.
```

Comment:
Users can override only node configuration and those shouldn't change between protocol upgrades.

Reply:
I think that should be possible, and maybe even necessary occasionally.  Say we add a new field to the config file, like a path to a second secret key - they'd have to provide that info.

Reply:
I didn't say we should not allow for overriding of node configuration but I see now what you meant by the original comment on "drawback" :+1: